### ISSUE TITLE
Support Multiple Currencies per Account

### ISSUE THREAD

**Issue by Mihail-K** [envato/double_entry#179]
Currently, an Account has a configured currency associated to it.
As I understand it, the currency is fixed and enforced for all transfers involving the account.
For our use-case, we'd like to be able to have a single account identifier (ie. `:cash`, `:vendor_payable`, etc.) to support multiple currencies.

To clarify, only a single currency would be involved in a transaction. 
For example:
```ruby
vendor = Vendor.take # Some vendor
order  = vendor.orders.payable.first # Some order from the vendor

cash           = DoubleEntry.account(:cash, currency: order.currency)
vendor_payable = DoubleEntry.account(:vendor_payable, currency: order.currency, scope: vendor)

DoubleEntry.transfer(order.vendor_total, from: cash, to: vendor_payable, code: :pay_vendor_for_order)
```
Essentially, `account(:cash, currency: 'CAD')` would hold a separate balance from `account(:cash, currency: 'USD')`.

In terms of technical changes required to accomplish this, it looks to be just a matter of adding a `currency` column to the `double_entry_lines` table, and scoping queries for Line records by currency (for calculating balances and the like).

While this could be accomplished by defining many different versions of each account (`:cash_cad`, `:cash_usd`, `:cash_sek`, `:cash_eur`, etc.), it would considerably complicate displaying breakdowns and balances to the internal and end users, and also force all possible transfers to be configured N times (once per currency).

**Comment by orien** [comment]
Hi @Mihail-K.

Thanks for your suggestion. I can understand how this would be a useful feature. However, there are no plans to implement this at the moment.

**Comment by ryano3000** [comment]
> Currently, an Account has a configured currency associated to it.
> As I understand it, the currency is fixed and enforced for all transfers involving the account.
> For our use-case, we'd like to be able to have a single account identifier (ie. `:cash`, `:vendor_payable`, etc.) to support multiple currencies.
> 
> To clarify, only a single currency would be involved in a transaction.
> For example:
> 
> ```ruby
> vendor = Vendor.take # Some vendor
> order  = vendor.orders.payable.first # Some order from the vendor
> 
> cash           = DoubleEntry.account(:cash, currency: order.currency)
> vendor_payable = DoubleEntry.account(:vendor_payable, currency: order.currency, scope: vendor)
> 
> DoubleEntry.transfer(order.vendor_total, from: cash, to: vendor_payable, code: :pay_vendor_for_order)
> ```
> 
> Essentially, `account(:cash, currency: 'CAD')` would hold a separate balance from `account(:cash, currency: 'USD')`.
> 
> In terms of technical changes required to accomplish this, it looks to be just a matter of adding a `currency` column to the `double_entry_lines` table, and scoping queries for Line records by currency (for calculating balances and the like).
> 
> While this could be accomplished by defining many different versions of each account (`:cash_cad`, `:cash_usd`, `:cash_sek`, `:cash_eur`, etc.), it would considerably complicate displaying breakdowns and balances to the internal and end users, and also force all possible transfers to be configured N times (once per currency).

I also need this feature - support multiple currencies without creating separate accounts. Did you find a solution @Mihail-k?

**Comment by ricobl** [comment]
Hey @ryano3000! Looks like Mihail was tagged here by mistake instead of @mintyfresh.

Unfortunately we still have no plans to support multiple currencies.

**Comment by ryano3000** [comment]
Thanks @ricobl 

I ended up implementing a multicurrency solution by simply having an array of currencies. Looping through it and dynamically defining account names using the currency as a suffix to the name. 

eg sales_account_usd 

I then wrote a small wrapper to fetch accounts scoped by currency 
eg.
DoubleEntryHelper.fetch_account(account_name, currency)
This then searches for the account scoped by currency (account_name_currency).

Working well so far!
